File: js/sdks/packages/tokenizer-sdk/src/tokenizer.ts
/*
 * Copyright 2021 by LunaSec (owned by Refinery Labs, Inc)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
import { OutgoingHttpHeaders } from 'http';

import { LunaSecError } from '@lunasec/isomorphic-common';
import axios, { AxiosError, AxiosResponse } from 'axios';
import { FileWithPath } from 'react-dropzone';

import { downloadFromS3WithSignedUrl, uploadToS3WithSignedUrl } from './aws';
import { AUTHORIZATION_HEADER, CONFIG_DEFAULTS, SESSION_HASH_HEADER } from './constants';
import { Configuration, DefaultApi, ErrorResponse, ErrorResponseError, MetaData } from './generated';
import {
  FileInfo,
  SuccessOrFailOutput,
  TokenizerClientConfig,
  TokenizerDetokenizeFileInfo,
  TokenizerDetokenizeFileResponse,
  TokenizerDetokenizeResponse,
  TokenizerDetokenizeToUrlResponse,
  TokenizerFailApiResponse,
  TokenizerGetMetadataResponse,
  TokenizerTokenizeResponse,
  TokenizerVerifyGrantResponse,
} from './types';

// Uses an openAPI generated client to query the tokenizer.  The biggest gotchas here are that:
// 1) Axios returns an res object with the tokenizer's response on the 'data' property, but the tokenizer also wraps its responses with 'data'
// So the actual response body is res.data.data.  We handle that here
// 2) Axios throws for any non 200 response code, and then we lose the typing of the error response, so we catch and then stick it back on.

export class Tokenizer {
  readonly config!: TokenizerClientConfig;
  readonly apiClient: DefaultApi;
  private reqOptions: { headers: OutgoingHttpHeaders };

  constructor(config: Partial<TokenizerClientConfig>) {
    // Deep clone config for mutation safety.
    this.config = JSON.parse(JSON.stringify(Object.assign({}, CONFIG_DEFAULTS, config)));

    const headers: OutgoingHttpHeaders = {
      'Content-Type': 'application/json',
    };
    const jwtToken = this.config.authenticationToken;
    if (jwtToken) {
      headers[AUTHORIZATION_HEADER] = jwtToken;
    }
    this.reqOptions = { headers }; // This is passed to the openapi client on every request

    // openapi stuff
    const openAPIConfig = new Configuration({ basePath: this.config.url });
    const axiosInstance = axios.create({});
    this.apiClient = new DefaultApi(openAPIConfig, undefined, axiosInstance);
  }

  private handleError(e: AxiosError | Error | unknown): TokenizerFailApiResponse {
    if (!(e instanceof Error)) {
      // this will likely never happen but TS makes you handle it
      throw e;
    }
    // this will be what happens
    return {
      success: false,
      error: this.constructError(e),
    };
  }

  private constructError(e: AxiosError<ErrorResponse> | Error): LunaSecError {
    if ('response' in e && e.response) {
      const errorWithCoercedType = e.response.data.error as ErrorResponseError & {
        errorMessage?: string;
      };

      // Parse the axios error, if it has any meaningful data about the response
      return new LunaSecError({
        name: e.response.data.error.name || 'TokenizerError',
        // TODO: Update this to "message" to conform with openAPI spec once the Tokenizer Backend uses OpenAPI
        message: errorWithCoercedType.message || errorWithCoercedType.errorMessage || 'Unknown Tokenizer Error',
        code: e.response.status.toString(),
      });
    }
    if (e instanceof Error) {
      return new LunaSecError(e); // This can handle axios errors where we dont even get a response, or any other case
    }
    return new LunaSecError({ name: 'unknownTokenizerError', message: 'Unknown Tokenization Error', code: '500' });
  }

  public async createGrant(sessionId: string, tokenId: string, customDuration?: string) {
    try {
      const res = await this.apiClient.setGrant(
        {
          sessionId,
          tokenId,
          customDuration,
        },
        this.reqOptions
      );
      return {
        success: res.data.success,
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  async verifyGrant(sessionId: string, tokenId: string): SuccessOrFailOutput<TokenizerVerifyGrantResponse> {
    try {
      const res = await this.apiClient.verifyGrant(
        {
          sessionId,
          tokenId,
        },
        this.reqOptions
      );
      return {
        success: true,
        valid: res.data.data.valid,
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  async getMetadata(tokenId: string): SuccessOrFailOutput<TokenizerGetMetadataResponse> {
    try {
      const res = await this.apiClient.getMetaData(
        {
          tokenId,
        },
        this.reqOptions
      );
      return {
        success: true,
        metadata: res.data.data.metadata,
        tokenId: tokenId, // Not sure why we pass this back, seems useless in this context
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  async tokenize(input: string | Buffer, metadata: MetaData): SuccessOrFailOutput<TokenizerTokenizeResponse> {
    try {
      const res = await this.apiClient.tokenize(
        {
          metadata,
        },
        this.reqOptions
      );
      const data = res.data.data;
      await uploadToS3WithSignedUrl(data.uploadUrl, data.headers as OutgoingHttpHeaders, input);

      return {
        success: true,
        tokenId: data.tokenId,
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  async detokenize(tokenId: string): SuccessOrFailOutput<TokenizerDetokenizeResponse> {
    const response = await this.detokenizeToUrl(tokenId);

    if (!response.success) {
      return response;
    }
    const { headers, downloadUrl } = response;
    return {
      success: true,
      tokenId: tokenId,
      value: await downloadFromS3WithSignedUrl(downloadUrl, headers),
    };
  }

  /**
   * Binds this instance of the tokenizer to a session so that any requests after the first detokenization will include
   * a hash generated by the Tokenizer Backend as a header.  Any requests that include a hash not matching the current session
   * will not be honored by the Backend, locking this tokenizer to the session of the first request it made
   */
  private handleSessionHashTracking(response: AxiosResponse): { success: false; error: LunaSecError } | null {
    if (!this.config.lockToSession) {
      return null;
    }
    // already bound to a session
    if (this.reqOptions.headers[SESSION_HASH_HEADER]) {
      return null;
    }
    const sessionHash = response.headers[SESSION_HASH_HEADER] as string | undefined;
    if (sessionHash === undefined) {
      return {
        success: false,
        error: new LunaSecError({
          name: 'detokenizationiFrameSessionBinding',
          message: 'session hash was not set in response when detokenizing, unable to bind iFrame to a session',
          code: '500',
        }),
      };
    }
    this.reqOptions.headers[SESSION_HASH_HEADER] = sessionHash;
    return null;
  }

  async detokenizeToUrl(tokenId: string): SuccessOrFailOutput<TokenizerDetokenizeToUrlResponse> {
    try {
      const response = await this.apiClient.detokenize(
        {
          tokenId,
        },
        this.reqOptions
      );
      const sessionError = this.handleSessionHashTracking(response);
      if (sessionError) {
        return sessionError;
      }

      const res = response.data;

      if (!res.data) {
        return {
          success: false,
          error: new LunaSecError({
            name: 'badDetokenizeResponse',
            message: 'Invalid response from Tokenizer when detokenizing data',
            code: '500',
          }),
        };
      }

      const { downloadUrl, headers } = res.data;
      return {
        success: true,
        tokenId: tokenId,
        headers: headers as Record<string, string>,
        downloadUrl: downloadUrl,
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  /**
   * // Fetches file info and pre-signed URL needed to start a file download.
   * Useful when we want to prepare the option to start a download but not yet start one.
   * Grabs the file URL and metadata in parallel.  Similar to detokenizeToUrl but for files
   * @param token
   */
  async detokenizeToFileInfo(token: string): SuccessOrFailOutput<TokenizerDetokenizeFileInfo> {
    const metaPromise = this.getMetadata(token);
    // TODO: make this function able to skip detokenizing to URL optionally because sometimes we dont need it.
    const urlPromise = this.detokenizeToUrl(token);
    const [metaRes, urlRes] = await Promise.all([metaPromise, urlPromise]);
    if (!metaRes.success) {
      return metaRes;
    }
    if (!urlRes.success) {
      return urlRes;
    }

    const meta = metaRes.metadata;

    if (meta.dataType !== 'file' || !('fileinfo' in meta)) {
      return {
        success: false,
        error: new LunaSecError({
          name: 'wrongMetaDataType',
          code: '400',
          message: "Couldn't find metadata information for a file, it may have been the wrong type of token.",
        }),
      };
    }

    const fileMeta = meta.fileinfo;
    const fileInfo: FileInfo = {
      filename: fileMeta.filename,
      options: {
        lastModified: fileMeta.lastModified,
        type: fileMeta.type,
      },
      headers: urlRes.headers,
      url: urlRes.downloadUrl,
    };
    return {
      success: true,
      fileInfo,
    };
  }

  /**
   * BROWSER ONLY
   * Triggers a download of a file that we already got the file info for using getFileInfo
   * @param fileInfo LunaSec's custom FileInfo object which tells us about the tokenized file
   */
  async detokenizeFileFromFileInfo(fileInfo: FileInfo): SuccessOrFailOutput<TokenizerDetokenizeFileResponse> {
    const axiosInstance = axios.create();
    try {
      const res = await axiosInstance.get(fileInfo.url, {
        headers: fileInfo.headers,
        responseType: 'blob',
      });
      return {
        success: true,
        file: new File([res.data], fileInfo.filename, fileInfo.options),
      };
    } catch (e) {
      return this.handleError(e);
    }
  }

  /**
   * BROWSER ONLY
   * Takes a token and downloads and returns a File object, complete with proper name, mime type, and lastModified fields
   * @param token
   */
  async detokenizeFile(token: string): SuccessOrFailOutput<TokenizerDetokenizeFileResponse> {
    const fileInfoRes = await this.detokenizeToFileInfo(token);
    if (!fileInfoRes.success) {
      return fileInfoRes;
    }
    return this.detokenizeFileFromFileInfo(fileInfoRes.fileInfo);
  }

  /**
   * BROWSER ONLY
   * uploads a file into LunaSec and returns a tokenId, just like `tokenize`
   * @param file
   * @param customMetadata
   */
  async tokenizeFile(
    file: File | FileWithPath,
    customMetadata?: Record<string, unknown>
  ): SuccessOrFailOutput<TokenizerTokenizeResponse> {
    try {
      const arrayBuf = await file.arrayBuffer();
      // Turn the JS ArrayBuffer into a Node type Buffer for tokenizer
      // todo: try skipping this
      const buf = Buffer.from(new Uint8Array(arrayBuf));
      const meta: MetaData = {
        dataType: 'file',
        fileinfo: {
          filename: file.name,
          type: file.type,
          lastModified: file.lastModified,
        },
        customFields: customMetadata,
      };
      return this.tokenize(buf, meta);
    } catch (e) {
      return this.handleError(e);
    }
  }
}

File: lunadefend/go/controller/tokenizer.go
// Copyright 2021 by LunaSec (owned by Refinery Labs, Inc)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package controller

import (
	"encoding/json"
	"github.com/lunasec-io/lunasec/lunadefend/go/constants"
	"github.com/lunasec-io/lunasec/lunadefend/go/util/auth"
	"io/ioutil"
	"log"
	"net/http"

	"go.uber.org/config"

	"github.com/lunasec-io/lunasec/lunadefend/go/service"
	"github.com/lunasec-io/lunasec/lunadefend/go/types"
	"github.com/lunasec-io/lunasec/lunadefend/go/types/event"
	"github.com/lunasec-io/lunasec/lunadefend/go/util"
	"github.com/pkg/errors"
)

type tokenizerController struct {
	tokenizerControllerConfig
	tokenizer   service.TokenizerService
	jwtVerifier service.JwtVerifier
	meta        service.MetadataService
	grant       service.GrantService
}

type tokenizerControllerConfig struct {
}

type TokenizerController interface {
	TokenizerGet(w http.ResponseWriter, req *http.Request)
	TokenizerSet(w http.ResponseWriter, req *http.Request)
}

func NewTokenizerController(
	provider config.Provider,
	tokenizer service.TokenizerService,
	jwtVerifier service.JwtVerifier,
	meta service.MetadataService,
	grant service.GrantService,
) (controller TokenizerController) {
	var (
		controllerConfig tokenizerControllerConfig
	)

	err := provider.Get("tokenizer_controller").Populate(&controllerConfig)
	if err != nil {
		panic(err)
	}

	controller = &tokenizerController{
		tokenizerControllerConfig: controllerConfig,
		tokenizer:                 tokenizer,
		jwtVerifier:               jwtVerifier,
		meta:                      meta,
		grant:                     grant,
	}
	return
}

func (s *tokenizerController) requestHasValidGrantForToken(r *http.Request, tokenID types.Token) (valid bool, err error) {
	claims, err := auth.GetRequestClaims(s.jwtVerifier, r)
	if err != nil {
		err = errors.Wrap(err, "unable to verify token jwt with claims")
		return
	}

	return s.grant.ValidTokenGrantExistsForSession(tokenID, claims.SessionID, constants.TokenFullAccess)
}

func (s *tokenizerController) TokenizerGet(w http.ResponseWriter, r *http.Request) {
	log.Printf("Received TokenizerGet request")

	input := event.TokenizerGetRequest{}
	b, err := ioutil.ReadAll(r.Body)

	if err != nil {
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	if err := json.Unmarshal(b, &input); err != nil {
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	valid, err := s.requestHasValidGrantForToken(r, types.Token(input.TokenID))
	if err != nil {
		log.Println(err)
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	if !valid {
		err = errors.New("session does not have valid token grant available to detokenize")
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	url, headers, err := s.tokenizer.TokenizerGet(types.Token(input.TokenID))
	if err != nil {
		statusCode := 500
		// TODO: Make this error message a constant
		if err.Error() == "unable to locate data for token" {
			statusCode = 404
		}
		util.RespondError(w, statusCode, err)
		return
	}

	resp := event.TokenizerGetResponse{
		DownloadURL: url,
		Headers:     headers,
	}

	util.Respond(w, resp)
}

func (s *tokenizerController) TokenizerSet(w http.ResponseWriter, r *http.Request) {
	log.Printf("Received TokenizerSet request")

	input := event.TokenizerSetRequest{}
	b, err := ioutil.ReadAll(r.Body)

	if err != nil {
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	if err := json.Unmarshal(b, &input); err != nil {
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	claims, err := auth.GetRequestClaims(s.jwtVerifier, r)
	if err != nil {
		err = errors.Wrap(err, "unable to verify token jwt with claims")
		util.RespondError(w, http.StatusBadRequest, err)
		return
	}

	tokenID, url, headers, err := s.tokenizer.TokenizerSet()
	if err != nil {
		util.RespondError(w, http.StatusInternalServerError, err)
		return
	}

	if len(input.Metadata) > 0 {
		if err := s.meta.SetMetadata(tokenID, claims, input.Metadata); err != nil {
			util.RespondError(w, http.StatusInternalServerError, err)
			return
		}
	}
	//We automatically create grants for new tokens so that they can be set in the DB, and detokenized elsewhere in the same browser session
	if err := s.grant.SetTokenGrantForSession(tokenID, claims.SessionID, constants.TokenFullAccess, ""); err != nil {
		util.RespondError(w, http.StatusInternalServerError, err)
		return
	}

	resp := event.TokenizerSetResponse{
		TokenID:   string(tokenID),
		UploadURL: url,
		Headers:   headers,
	}

	util.Respond(w, resp)
}
